### PR TITLE
[IMP] mail: improved composer actions (+)

### PR DESCRIPTION
**Pull request by alexkuhn**
This commit adds a "+" dropup menu to the Discuss composer, whether in discuss app or chat window, on the left side of the composer.

This "+" button contains secondary features of the composer, which helps reducing the amount of items in composer that are immediately available. This also gives the possibility to add more features in composer without being limited to the amount of quick actions, especially on narrow views of the conversation like in chat window.

This PR also makes the other related improvements:
1. Composer actions are defined in a registry, similarly to how thread and message actions are defined
2. Component "Picker" has been removed and instead its mobile nice view is built-in to all `useEmojiPicker()` users.
3. Composer actions buttons are bigger so more visible.
4. Send quick button has improved style, making it more obvious when composer has content that can be sent. This is also easier to use on mobile devices
5. The many tests in Discuss posting messages are now using `press("Enter")`. This make tests DRyer and matches with the common interaction to post messages in Discuss.

Task-4210591
Task-4311296

https://github.com/odoo/enterprise/pull/73346

Screenshot 2024-11-08 at 19 32 36

### CODE REVIEW

File: addons/mail/static/src/core/common/message_actions.js [hunk 1]
```diff
@@ -7,72 +7,42 @@ import { MessageReactionButton } from "./message_reaction_button";
...
-                        );
-                        if (!reaction) {
-                            component.props.message.react(emoji);
-                        }
-                        def.resolve(true);
-                    },
+        onClick: async (component, action) => {
+            return component.reactionPicker.open({
+                el: component.root?.el?.querySelector(`[name="${action.id}"]`),
+            });
```
Comment (tsm-odoo) [inline addons/mail/static/src/core/common/message_actions.js:30]: Could remove curly brackets and return, would be shorter and easier to read.

File: addons/mail/static/src/core/common/composer.xml [hunk 2]
```diff
@@ -2,24 +2,33 @@
 <templates xml:space="preserve">
 
 <t t-name="mail.Composer">
+    <t t-set="isChatterComposer" t-value="extended and !props.composer.message"/>
+    <t t-set="partitionedActions" t-value="composerActions.partition"/>
     <t t-set="compact" t-value="props.mode === 'compact'"/>
     <t t-set="normal" t-value="props.mode === 'normal'"/>
     <t t-set="extended" t-value="props.mode === 'extended'"/>
-    <div t-ref="composer">
+    <t t-set="actionsContainerClass" t-value="'o-mail-Composer-actions d-flex rounded-3'"/>
```
Comment (phenix-factory) [inline addons/mail/static/src/core/common/composer.xml:10]: I have never been a fan of `t-set`, can't those be getters? It's easier if we need to override them in the future. 

Also, the `partionedActions` look useless?

Comment (alexkuhn) [reply addons/mail/static/src/core/common/composer.xml:10]: `partitionedActions` is to memoize the result from `composerActions.partition`

Comment (alexkuhn) [reply addons/mail/static/src/core/common/composer.xml:10]: `actionContainerClass` being a `t-set` has the advantage to be immediately visible in code of template, so easier to grep there. Getter is more powerful but it's annoying without single-file component...

So there's pros and cons with both approaches. Don't have strong stance either way although slight preference for `t-set` when this is enough.

File: addons/mail/static/src/core/common/composer.scss [hunk 3]
```diff
@@ -112,7 +147,11 @@
     }
 
     &::placeholder {
-        opacity: 75%;
+        opacity: 40%;
+
+        .o-mail-Composer.o-focused & {
+            opacity: 50%;
```
Comment (phenix-factory) [inline addons/mail/static/src/core/common/composer.scss:153]: 10% opacity is barely a change, I suggest we use the initial 75%

File: addons/mail/static/src/core/common/composer.js [hunk 4]
```diff
@@ -713,7 +705,9 @@ export class Composer extends Component {
         const secondPart = text.slice(composer.selection.end, text.length);
         composer.text = firstPart + str + secondPart;
         this.selection.moveCursor((firstPart + str).length);
-        if (!this.ui.isSmall) {
+        if (this.ui.isSmall) {
+            return false;
+        } else {
```
Comment (phenix-factory) [inline addons/mail/static/src/core/common/composer.js:710]: I don't get this change, what's the point of this `return false`?

Comment (alexkuhn) [reply addons/mail/static/src/core/common/composer.js:710]: It's to not close the picker in mobile. This makes it more natural to use picker in mobile and matches the behaviour of our competitors.

File: addons/mail/static/src/core/common/composer.scss (hunk 3 above)
Comment (alexkuhn) [reply addons/mail/static/src/core/common/composer.scss:153]: This is relatively a 20%, so still visible of a change.
I'd not increase more than 50%, the placeholder is too visible and looks like composed text. This makes using composer more exhausting than it should.

File: addons/mail/static/src/core/common/composer.js (hunk 4 above)
Comment (phenix-factory) [reply addons/mail/static/src/core/common/composer.js:710]: It was already the case before no? This `false` doesn't look useful when I read the code.

**Comment by alexkuhn** [comment]
@robodoo r+

**Pull request closed without merging**